### PR TITLE
feat: re-assign person faces (2)

### PR DESCRIPTION
**Pull request by martabal**
## Changes made in this PR

This PR introduces the ability to re-assign an asset from a person. The asset can be assigned with either a new or an existing person.
With all the recent changes, I changed a bit the global logic. The detail panel now displays people detected in the asset. while the new modal (person-side-panel) is designed to handle faces in the asset. With these changes, you can re-assign twins detected as the same person in the same asset. I added a rectangle on people faces when hovering image thumbnails (resolves #3489).

This PR closes #4153 and global edit has been added to this PR ~global edit will come in a future PR~.

Future PRs:
- The third PR will add the ability to assign and un-assign faces.
- The fourth PR will add the ability to manually tag faces within an image.

### Re-assigning faces

https://github.com/immich-app/immich/assets/74269598/84c63fc7-9b6e-44f2-8798-88dd1dcec898

### Hovering thumbnails

https://github.com/immich-app/immich/assets/74269598/c2efd9f8-8efc-43e6-afc8-d24ff4acc76c

### CODE REVIEW

**Comment by martabal** [comment]
I transitioned from an optimistic approach involving people and faces, encountering errors and edge-cases on complex flows. Now,it uses websocket: when faces are only reassigned, the client seamlessly pulls updated asset information. In cases where faces are re-assigned to new people, the client waits for the server to generate all the new thumbnails before going back to the detail panel.

File: server/src/domain/access/access.core.ts [hunk 1]
```diff
@@ -244,6 +246,11 @@ export class AccessCore {
       case Permission.PERSON_MERGE:
         return this.repository.person.hasOwnerAccess(authUser.id, id);
 
+      case Permission.PERSON_CREATE:
```
Comment (jrasm91) [inline server/src/domain/access/access.core.ts:249]: It looks like this is really about updating a face, not creating a person.

File: server/src/domain/person/person.dto.ts [hunk 2]
```diff
@@ -73,6 +73,53 @@ export class PersonResponseDto {
...
+  @IsArray()
+  @ValidateNested({ each: true })
+  @Type(() => AssetFaceUpdateItem)
+  data!: AssetFaceUpdateItem[];
+}
+
+export class AssetFaceUpdateItem {
+  @Optional()
+  @IsString()
+  assetFaceId?: string;
```
Comment (jrasm91) [inline server/src/domain/person/person.dto.ts:112]: Should this just be called id?

File: server/src/domain/person/person.service.ts [hunk 3]
```diff
@@ -80,6 +83,104 @@ export class PersonService {
     };
   }
 
+  async createPerson(authUser: AuthUserDto, dto: AssetFaceUpdateDto): Promise<PersonResponseDto> {
+    const changeFeaturePhoto: string[] = [];
+
+    const newPerson = await this.repository.create({ ownerId: authUser.id });
```
Comment (jrasm91) [inline server/src/domain/person/person.service.ts:89]: It is possible to create a new person without the "person_create" permission. It looks like it is possible to create a person without actually linking it to anything.

Should we do the linking here in the server? Would it be better for the client to call a separate endpoint to re-assign a face to the newly created person? Sometimes when the server touches lots of resources like this it becomes not very restful and makes it a bit overly complicated IMO.

**Review by jrasm91 — COMMENTED** [review]
To me, the way these endpoints are setup is quite confusing right now. Can we have something that is more restful?

### Person
- `POST /person` (create)
- `GET /person/id` (get by id)
- `PUT /person/id` (update)
- `DELETE /person/id` (delete)
### Face
- `GET /faces?personId=x&assetId=y` (get by personId or assetId)
- `PUT /faces/id` (update)

Based on some of the code you have added, you have almost created a full CRUD implementation for faces, independently of `/person`. It might be clearer to add a face controller (in the same person.controller.ts file is fine) that exposes the new features with alternative endpoints.

IMO we definitely don't need two implementations or re-assign in the person.service codebase though.

File: server/src/immich/controllers/person.controller.ts [hunk 4]
```diff
@@ -34,6 +36,25 @@ export class PersonController {
...
+    return this.service.reassignFaces(authUser, id, dto);
+  }
+
+  @Post('')
+  createPerson(@AuthUser() authUser: AuthUserDto, @Body() dto: AssetFaceUpdateDto): Promise<PersonResponseDto> {
+    return this.service.createPerson(authUser, dto);
+  }
+
+  @Get('faces/:id')
+  getFaces(@AuthUser() authUser: AuthUserDto, @Param() { id }: UUIDParamDto): Promise<AssetFaceResponseDto[]> {
```
Comment (jrasm91) [inline server/src/immich/controllers/person.controller.ts:54]: Where is this endpoint used?

**Comment by alextran1502** [comment]
@jrasm91, do you have anything else to add about the server code?

File: server/src/domain/person/person.dto.ts [hunk 5]
```diff
@@ -73,6 +73,47 @@ export class PersonResponseDto {
...
 }
 
+export class PersonWithFacesResponseDto extends PersonResponseDto {
+  faces!: AssetFaceWithoutPersonResponseDto[];
+}
+
+export class AssetFaceWithoutPersonResponseDto {
+  @IsString()
+  @IsNotEmpty()
+  id!: string;
```
Comment (jrasm91) [inline server/src/domain/person/person.dto.ts:83]: This should use `@ValidateUUID()`

File: server/src/domain/person/person.dto.ts [hunk 6]
```diff
@@ -73,6 +73,47 @@ export class PersonResponseDto {
...
+  @Type(() => AssetFaceUpdateItem)
+  data!: AssetFaceUpdateItem[];
+}
+
+export class AssetFaceUpdateItem {
+  @IsString()
+  personId!: string;
+
+  @IsString()
+  assetId!: string;
```
Comment (jrasm91) [inline server/src/domain/person/person.dto.ts:114]: These should also both be uuids.

File: server/src/domain/face/face.dto.ts
Comment (jrasm91) [inline server/src/domain/face/face.dto.ts:1]: Can you just move this into the person dto file?

File: server/src/domain/search/search.service.ts [hunk 7]
```diff
@@ -159,7 +159,7 @@ export class SearchService {
     };
   }
 
-  searchPerson(authUser: AuthUserDto, dto: SearchPeopleDto): Promise<PersonResponseDto[]> {
+  async searchPerson(authUser: AuthUserDto, dto: SearchPeopleDto): Promise<PersonResponseDto[]> {
```
Comment (jrasm91) [inline server/src/domain/search/search.service.ts:162]: This `async` isn't needed here.

File: server/src/immich/controllers/face.controller.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    @Param() { id }: UUIDParamDto,
+    @Body() dto: FaceDto,
+  ): Promise<PersonResponseDto> {
+    return this.service.reassignFacesById(authUser, id, dto);
+  }
+
+  @Get()
+  getFaces(@AuthUser() authUser: AuthUserDto, @Query() dto: FaceDto): Promise<AssetFaceResponseDto[]> {
+    return this.service.getFacesById(authUser, dto);
+  }
```
Comment (jrasm91) [inline server/src/immich/controllers/face.controller.ts:28]: These should probably be swapped, the get and then the put (CRUD)

File: server/src/immich/controllers/person.controller.ts [hunk 9]
```diff
@@ -34,6 +35,20 @@ export class PersonController {
...
+    @Param() { id }: UUIDParamDto,
+    @Body() dto: AssetFaceUpdateDto,
+  ): Promise<PersonResponseDto[]> {
+    return this.service.reassignFaces(authUser, id, dto);
+  }
+
+  @Post()
+  createPerson(@AuthUser() authUser: AuthUserDto): Promise<PersonResponseDto> {
+    return this.service.createPerson(authUser);
+  }
```
Comment (jrasm91) [inline server/src/immich/controllers/person.controller.ts:50]: swap these ones too create before update.

**Review by jrasm91 — APPROVED** [review]
Looks good for the most part, just a few nits.

File: web/src/lib/stores/people.store.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+  imageWidth: number;
+  boundingBoxX1: number;
+  boundingBoxX2: number;
+  boundingBoxY1: number;
+  boundingBoxY2: number;
+}
+
+export const boundingBoxesArray = writable<Faces[]>([]);
+
+export const setBoundingBoxesArray = (value: Faces[]) => {
```
Comment (alextran1502) [inline web/src/lib/stores/people.store.ts:14]: Should we just set the `boundingBoxesArray` directly in the component?

File: web/src/lib/components/asset-viewer/photo-viewer.svelte [hunk 11]
```diff
@@ -129,12 +134,19 @@
...
     <div bind:this={imgElement} class="h-full w-full">
       <img
+        bind:this={$photoViewer}
         transition:fade={{ duration: haveFadeTransition ? 150 : 0 }}
         src={assetData}
         alt={asset.id}
         class="h-full w-full object-contain"
         draggable="false"
       />
+      {#each showBoundingBox($boundingBoxesArray, $photoZoomState, $photoViewer) as boundingbox}
```
Comment (alextran1502) [inline web/src/lib/components/asset-viewer/photo-viewer.svelte:144]: Maybe `getBoundingBox` instead?

**Comment by Merlin-2504** [comment]
I assigned faces incorrectly. Does the method from the video still work? I followed the steps in the video, but it isn't working.

**Pull request merged**